### ISSUE TITLE
Allow inputs to be consumed as `Provider`/`Optional`

### ISSUE THREAD

**Issue by lacasseio** [nokeedev/gradle-native#448]
We may want to trigger on some specific inputs only once either when 1) they first appear or 2) before they appear.

In the first scenario, if the input never comes to exist, the action is never executed. It's important to note that the deferred type will contain a value, but the value may change in the future so delaying the resolution of the deferred input type is important in order to get the most up-to-date value.

In the second scenario, the action is executed regardless if the input exists or not. It's up to the user to deal with the existence or not of the input. It's important to note that the input may exist in the future and realizing the deferred input type would give the value at its present state (maybe null or not).

In both cases, the realization of the deferred input type can have a cascading effect on the state of the objects, such as forcing values to finalize hence making them immutable. There is nothing the universal model can do to work around those side effect outside of tracking and reporting them.
